File: 065_EditDistance.cpp
// Topic: Dynamic Programming
// Edit Distance: classic Levenshtein distance (insert/delete/replace)
// Time Complexity: O(m*n)
// Space Complexity: O(min(m,n)) optimized

#include <bits/stdc++.h>
using namespace std;

int minDistance(string word1, string word2)
{
    if (word1.size() < word2.size())
        swap(word1, word2);
    int m = word1.size(), n = word2.size();
    vector<int> dp(n + 1);
    iota(dp.begin(), dp.end(), 0);
    for (int i = 1; i <= m; i++)
    {
        vector<int> ndp(n + 1);
        ndp[0] = i;
        for (int j = 1; j <= n; j++)
        {
            if (word1[i - 1] == word2[j - 1])
                ndp[j] = dp[j - 1];
            else
                ndp[j] = 1 + min({dp[j], ndp[j - 1], dp[j - 1]});
        }
        dp.swap(ndp);
    }
    return dp[n];
}

int main()
{
    cout << minDistance("horse", "ros") << "\n";            // 3
    cout << minDistance("intention", "execution") << "\n";  // 5
    return 0;
}

File: 065_EditDistance.py
"""
Topic: Dynamic Programming
Problem: Edit Distance
Description: Minimum operations (insert/delete/replace) to convert word1 to word2.
Time: O(n*m)
Space: O(min(n,m)) optimized
"""

from typing import List


def minDistance(word1: str, word2: str) -> int:
    if len(word1) < len(word2):
        word1, word2 = word2, word1
    m, n = len(word1), len(word2)
    prev = list(range(n+1))
    for i in range(1, m+1):
        cur = [i] * (n+1)
        for j in range(1, n+1):
            if word1[i-1] == word2[j-1]:
                cur[j] = prev[j-1]
            else:
                cur[j] = 1 + min(prev[j], cur[j-1], prev[j-1])
        prev = cur
    return prev[n]


if __name__ == '__main__':
    print(minDistance("horse", "ros"))  # 3
    print(minDistance("intention", "execution"))  # 5
